  # Capstone Project Notebook #

This notebook will primarily be used for the capstone project as part of the Coursera IBM Data Science Course

In [16]:
import pandas as pd
import numpy as np
import requests

## Getting Wiki Data & Cleaning
1. Removing rows that do not have an assigned borough
2. If a cell has a borough but is not assigned a neighborhood, then we set the neighborhood to be the same as the borough (note that this condition is never met in the dataset, so we omit the operation).

In [5]:
df = pd.read_html("https://en.wikipedia.org/w/index.php?title=List_of_postal_codes_of_Canada:_M&oldid=1011037969", attrs={"class": "wikitable sortable"})[0]
df = df[df['Borough']!='Not assigned'].reset_index(drop=True)
df.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [6]:
df.shape

(103, 3)

## Merging with Geospatial Data
1. Given an external datatset containing latitude and longitude geospatial coordinates for each Postal Code, which we denote below as geo_df, we merge the two dataframes

In [7]:
geo_df = pd.read_csv("~/Desktop/Geospatial_Coordinates.csv")
merged_df = pd.merge(df,geo_df, on='Postal Code')
merged_df.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [128]:
print(merged_df.shape)

(103, 5)


We can be confident that the merging went smoothly (without losing data), because the dataframe shapes are consistent with expectations

## Visualising Latitude & Longitude Data

In [9]:
!conda install -c conda-forge folium=0.5.0 --yes -q
import folium

Solving environment: ...working... done

# All requested packages already installed.



In [10]:
latitude, longitude = [43.741667,-79.373333] ## found manually on wiki page

# create map of Toronto, Canada using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighbourhood in zip(merged_df['Latitude'], merged_df['Longitude'], merged_df['Borough'], merged_df['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
         [lat, lng],
         radius=5,
         popup=label,
         color='blue',
         fill=True,
         fill_color='#3186cc',
         fill_opacity=0.7,
         parse_html=False).add_to(map_toronto)  
    
map_toronto

## Segmenting Toronto Boroughs

#### Define FourSquare Credentials & Version

In [19]:
CLIENT_ID = 'NCOK3XLOBXPSZAGSGJ5ND3F0MOYUQGHQIMKHXHWFVSOUZWYV' # your Foursquare ID
CLIENT_SECRET = 'UPPMLT05HXXFDSVBYTEKTPRIBRRHFBMHB35REQM1QH2XPI5B' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value
radius=500

In [17]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [18]:
toronto_venues = getNearbyVenues(names=merged_df['Neighbourhood'],
                                   latitudes=merged_df['Latitude'],
                                   longitudes=merged_df['Longitude']
                                  );

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [134]:
toronto_venues.groupby('Neighbourhood').size()

Neighbourhood
Agincourt                                           5
Alderwood, Long Branch                              9
Bathurst Manor, Wilson Heights, Downsview North    21
Bayview Village                                     4
Bedford Park, Lawrence Manor East                  24
                                                   ..
Willowdale, Willowdale West                         5
Woburn                                              3
Woodbine Heights                                    5
York Mills West                                     3
York Mills, Silver Hills                            1
Length: 95, dtype: int64

### When we run groupby, neighbourhoods that don't return any venues on FourSquare are being omitted, I think

merged_df has shape (103,...) while toronto_venues.groupby('Neighbourhood').size has shape (95,...)

Therefore, these neighbourhoods contribute nothing to the clustering - when we append kmeans labels, we should remove neighbourhoods that are given a label of NaN.

In [135]:
toronto_venues.groupby('Neighbourhood').count()

Neighborhood Latitude  \
Neighbourhood                                                            
Agincourt                                                            5   
Alderwood, Long Branch                                               9   
Bathurst Manor, Wilson Heights, Downsview North                     21   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   24   
...                                                                ...   
Willowdale, Willowdale West                                          5   
Woburn                                                               3   
Woodbine Heights                                                     5   
York Mills West                                                      3   
York Mills, Silver Hills                                             1   

                                                 Neighborhood Longitude  \
Neighbourhood                                                             
Agincourt                                                             5   
Alderwood, Long Branch                                                9   
Bathurst Manor, Wilson Heights, Downsview North                      21   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    24   
...                                                                 ...   
Willowdale, Willowdale West                                           5   
Woburn                                                                3   
Woodbine Heights                                                      5   
York Mills West                                                       3   
York Mills, Silver Hills                                              1   

                                                 Venue  Venue Latitude  \
Neighbourhood                                                            
Agincourt                                            5               5   
Alderwood, Long Branch                               9               9   
Bathurst Manor, Wilson Heights, Downsview North     21              21   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   24              24   
...                                                ...             ...   
Willowdale, Willowdale West                          5               5   
Woburn                                               3               3   
Woodbine Heights                                     5               5   
York Mills West                                      3               3   
York Mills, Silver Hills                             1               1   

                                                 Venue Longitude  \
Neighbourhood                                                      
Agincourt                                                      5   
Alderwood, Long Branch                                         9   
Bathurst Manor, Wilson Heights, Downsview North               21   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             24   
...                                                          ...   
Willowdale, Willowdale West                                    5   
Woburn                                                         3   
Woodbine Heights                                               5   
York Mills West                                                3   
York Mills, Silver Hills                                       1   

                                                 Venue Category  
Neighbourhood                                                    
Agincourt                                                     5  
Alderwood, Long Branch                                        9 

In [136]:
print('There are {} unique categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 271 unique categories.


In [137]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]


In [138]:
toronto_onehot.shape

(2103, 272)

In [139]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Accessories Store  \
0                                         Agincourt                0.0   
1                            Alderwood, Long Branch                0.0   
2   Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                   Bayview Village                0.0   
4                 Bedford Park, Lawrence Manor East                0.0   
..                                              ...                ...   
90                      Willowdale, Willowdale West                0.0   
91                                           Woburn                0.0   
92                                 Woodbine Heights                0.0   
93                                  York Mills West                0.0   
94                         York Mills, Silver Hills                0.0   

    Airport  Airport Food Court  Airport Lounge  Airport Service  \
0       0.0                 0.0             0.0              0.0   
1       0.0                 0.0             0.0              0.0   
2       0.0                 0.0             0.0              0.0   
3       0.0                 0.0             0.0              0.0   
4       0.0                 0.0             0.0              0.0   
..      ...                 ...             ...              ...   
90      0.0                 0.0             0.0              0.0   
91      0.0                 0.0             0.0              0.0   
92      0.0                 0.0             0.0              0.0   
93      0.0                 0.0             0.0              0.0   
94      0.0                 0.0             0.0              0.0   

    Airport Terminal  American Restaurant  Antique Shop  Aquarium  ...  \
0                0.0             0.000000           0.0       0.0  ...   
1                0.0             0.000000           0.0       0.0  ...   
2                0.0             0.000000           0.0       0.0  ...   
3                0.0             0.000000           0.0       0.0  ...   
4                0.0             0.041667           0.0       0.0  ...   
..               ...                  ...           ...       ...  ...   
90               0.0             0.000000           0.0       0.0  ...   
91               0.0             0.000000           0.0       0.0  ...   
92               0.0             0.000000           0.0       0.0  ...   
93               0.0             0.000000           0.0       0.0  ...   
94               0.0             0.000000           0.0       0.0  ...   

    Train Station  Truck Stop  Vegetarian / Vegan Restaurant  \
0             0.0         0.0                            0.0   
1             0.0         0.0                            0.0   
2             0.0         0.0                            0.0   
3             0.0         0.0                            0.0   
4             0.0         0.0                            0.0   
..            ...         ...                            ...   
90            0.0         0.0                            0.0   
91            0.0         0.0                            0.0   
92            0.0         0.0                            0.0   
93            0.0         0.0                            0.0   
94            0.0         0.0                            0.0   

    Video Game Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  \
0                0.0                    0.0              0.0       0.0   
1                0.0                    0.0              0.0       0.0   
2                0.0                    0.0              0.0       0.0   
3                0.0                    0.0              0.0       0.0   
4                0.0                    0.0              0.0       0.0   
..               ...                    ...              ...       ...   
90               0.0                    0.0              0.0       0.0   
91               0.0                    0.0              0.0       0.0   
92      

In [140]:
toronto_grouped.shape

(95, 272)

In [141]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0             Clothing Store   0.2
1                     Lounge   0.2
2             Breakfast Spot   0.2
3  Latin American Restaurant   0.2
4               Skating Rink   0.2


----Alderwood, Long Branch----
                venue  freq
0         Pizza Place  0.22
1        Skating Rink  0.11
2      Sandwich Place  0.11
3                 Pub  0.11
4  Athletics & Sports  0.11


----Bathurst Manor, Wilson Heights, Downsview North----
         venue  freq
0  Coffee Shop  0.10
1         Bank  0.10
2  Pizza Place  0.05
3  Gas Station  0.05
4     Pharmacy  0.05


----Bayview Village----
                 venue  freq
0                 Bank  0.25
1  Japanese Restaurant  0.25
2                 Café  0.25
3   Chinese Restaurant  0.25
4  Moroccan Restaurant  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0  Italian Restaurant  0.08
1      Sandwich Place  0.08
2         Coffee Shop  0.08
3   Indian Restaurant  0.04
4  

                venue  freq
0                Park  0.33
1         Swim School  0.33
2            Bus Line  0.33
3   Accessories Store  0.00
4  Mexican Restaurant  0.00


----Leaside----
                    venue  freq
0             Coffee Shop  0.09
1     Sporting Goods Shop  0.09
2  Furniture / Home Store  0.06
3            Burger Joint  0.06
4                    Bank  0.06


----Little Portugal, Trinity----
                   venue  freq
0                    Bar  0.14
1            Men's Store  0.05
2  Vietnamese Restaurant  0.05
3             Restaurant  0.05
4            Coffee Shop  0.05


----Malvern, Rouge----
                             venue  freq
0             Fast Food Restaurant   1.0
1                    Metro Station   0.0
2  Molecular Gastronomy Restaurant   0.0
3       Modern European Restaurant   0.0
4                Mobile Phone Shop   0.0


----Milliken, Agincourt North, Steeles East, L'Amoreaux East----
                       venue  freq
0               Intersection

                             venue  freq
0                             Park   1.0
1                Accessories Store   0.0
2                    Metro Station   0.0
3  Molecular Gastronomy Restaurant   0.0
4       Modern European Restaurant   0.0




In [142]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [143]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

#neighbourhoods_venues_sorted.head()
neighbourhoods_venues_sorted.shape

(95, 11)

In [144]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([4, 0, 4, 4, 4, 4, 4, 4, 4, 4, 1, 4, 4, 4, 4, 4, 0, 4, 4, 4, 4, 4,
       4, 4, 4, 4, 0, 0, 4, 4, 4, 4, 0, 4, 4, 4, 4, 4, 0, 3, 4, 4, 4, 4,
       0, 4, 4, 4, 4, 2, 4, 4, 4, 4, 4, 4, 4, 4, 4, 0, 4, 4, 4, 4, 1, 4,
       4, 4, 0, 0, 0, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 0, 4, 0, 4,
       1, 4, 0, 4, 4, 1, 1], dtype=int32)

In [147]:
# add clustering labels
try:
    neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
except ValueError:
    pass

toronto_merged = merged_df

## toronto_merged needs to have elements removed if the clustering label is NaN.
toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')
#toronto_merged = toronto_merged.astype({"Cluster Labels": int})

# add clustering labels
#neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
###toronto_merged = merged_df

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
###toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')
#toronto_merged = toronto_merged.astype({"Cluster Labels": int})
#toronto_merged.dtypes

toronto_merged["Cluster Labels"].unique()


#toronto_merged.head()

array([ 4.,  0., nan,  2.,  1.,  3.])

In [70]:
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=10)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

TypeError: list indices must be integers or slices, not float